FEATURE: node:repair: remove nodes with invalid dimension values

robertlemke opened the pull request:
This change contains an improvement for node:repair which runs an
additional check for nodes which have dimension values which are not
allowed according to the current dimension configuration.

These nodes can be either removed or the user can decide to migrate them
by other means (for example a node migration).

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +861,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+     * @param boolean $dryRun Simulate?
+     * @return void
+     */
+    public function removeNodesWithInvalidDimensions(NodeType $nodeType = null, $workspaceName, $dryRun)
```
optional before required arguments?

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +861,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+        } else {
+            $this->output->outputLine('Checking for nodes with invalid dimensions ...');
+            $nodesArray = [];
+            foreach ($this->nodeTypeManager->getNodeTypes() as $nodeType) {
```
That seems a bit complicated, can we not simply grab all nodes regardless of nodeType in this case?

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +861,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+            }
+        }
+
+        if (count($nodesArray) === 0) {
```
checking for empty array is usually faster with `$nodesArray === []` :)

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +861,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+                continue;
+            }
+            foreach ($allowedDimensionCombinations as $dimensionConfiguration) {
+                if ($dimensionConfiguration == $nodeDataArray['dimensionValues']) {
```
=== ?

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +861,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+            ->setParameter('removed', false, \PDO::PARAM_BOOL);
+
+        foreach ($queryBuilder->getQuery()->getArrayResult() as $nodeDataArray) {
+            if ($nodeDataArray['path'] === '/') {
```
Mmmm, what about `/sites` ? And in the future maybe other node trees that do not use the dimensions?
Sounds dangerous ;)

robertlemke replied:
right. Maybe just skip if there are no dimension values at all?

kitsunet replied:
Sounds like an idea.

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +861,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+            if ($nodeDataArray['path'] === '/') {
+                continue;
+            }
+            foreach ($allowedDimensionCombinations as $dimensionConfiguration) {
```
Maybe it would be nicer to do one query per allowed configuration? not sure...

kitsunet commented:
Left a few comments.

robertlemke commented:
Okay, just tested this again, I feel confident that it works - at least I tried it with one Neos setup which had the issue of invalid dimension nodes and I could fix it.

kitsunet commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -856,6 +865,95 @@ protected function removeNode($nodeIdentifier, $dimensionsHash)
+        if ($nodeType !== null) {
+            $this->output->outputLine('Checking nodes of type "%s" for invalid dimensions ...', array($nodeType->getName()));
+            $nodesArray = $this->collectNodesWithInvalidDimensions($nodeType, $workspaceName, $allowedDimensionCombinations);
+        } else {
```
Just realized this case doesn't exist, as the signature of the method doesn't even allow giving `null` for the $nodeType... IMHO either we throw this code away or rework the signature. I mean actually you shouldn't need a node type to remove nodes with invalid dimensions. And looking at the call it seems NodeType can be `null` so the signature has to be reworked to make nodeType an optional argument.

robertlemke replied:
right, I'll fix that - also just found another oddity.

kitsunet commented:
Left one comment, apart from that looks good to me.

robertlemke commented:
@kitsunet okay, I still found a bug 😊  - but I also fixed it. The `break 2` clause was stupid. Because of it, there were still invalid dimension nodes left when method was done.

With the new implementation I can't find any invalid dimension nodes anymore when I manually search for it in the database - so it seems like it's working fine now.

I also took your advice and removed the node type specific code. Much better now.

Cross-checked this twice now with real data. I feel confident now.

aertmann commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -392,7 +397,11 @@ public function addMissingDefaultValuesByNodeType(NodeType $nodeType = null, $wo
+                    if ($node->getNodeData()->getDimensionValues() === []) {
+                        $this->output->outputLine('Skipping node %s  because it has no dimension values set', [$node->getPath()]);
+                    } else {
+                        $this->output->outputLine('Skipping node %s  because it has invalid dimension values: %s', [$node->getPath(), json_encode($node->getNodeData()->getDimensionValues())]);
```
same here

aertmann commented on TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Command/NodeCommandControllerPlugin.php:
```diff
@@ -392,7 +397,11 @@ public function addMissingDefaultValuesByNodeType(NodeType $nodeType = null, $wo
                 if ($node instanceof Node && !$node->dimensionsAreMatchingTargetDimensionValues()) {
-                    $this->output->outputLine('Skipping node %s  because it has invalid dimension values: %s', [$node->getPath(), json_encode($node->getNodeData()->getDimensionValues())]);
+                    if ($node->getNodeData()->getDimensionValues() === []) {
+                        $this->output->outputLine('Skipping node %s  because it has no dimension values set', [$node->getPath()]);
```
double space before because..

aertmann commented:
looks good from the code, found minor two typos

robertlemke commented:
Thanks for the review @aertmann! Also fixed the typo now.
